### PR TITLE
Support Framing for Window Aggregations

### PR DESCRIPTION
**Pull request by imply-cheddar**
This adds support for framing  over ROWS for window aggregations.

Still not implemented as yet:
1. RANGE frames
2. Multiple different frames in the same query
3. Frames on last/first functions

This PR also brings with it some adjustments and improvements to the tests for semantic interfaces and RowsAndColumns implementations.

The PR touches a lot of files, that's mostly because I combined all of the "semantic interfaces" into a `semantic` package and that affected a bunch of imports.

The PR also looks like it's a lot of lines of code, but most of that is actually whitespace changes on some of the test files done by auto-formatting.  Checking them in now so that future auto-formatting is safe from generating whitespace changes.

The primary code to review is essentially all inside of FramedOnHeapAggregatable, the tests associated with it and some in the planning.

Lastly, there are some code cleanup/renames being included that were asked for in the previous PR.  (Things like completing the renaming of `cell` and removing the TODOs)

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+        return computeReverseCumulativeAggregates(aggFactories, frame.getLowerOffset());
+      } else {
+        final int numRows = rac.numRows();
+        int lowerOffset = frame.getLowerOffset();
+        int upperOffset = frame.getUpperOffset();
+
+        if (numRows < lowerOffset + upperOffset + 1) {
+          // In this case, there are not enough rows to completely build up the full window aperture before it needs to
+          // also start contracting the aperture because of the upper offset. So we use a method that specifically
+          // handles checks for both expanding and reducing the aperture on every iteration.
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:77]: just wanted to say thanks for all the comments in this class, a lot going on here and it helps out :+1:

**Review by clintropolis — APPROVED** [review]
:rocket:

File: processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+
+public class WindowFrame
+{
+  @SuppressWarnings("unused")
+  public enum PeerType
```
[the remark this answers is not in the document]
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java:30]: Yeah, word.  This is a SQL concept stolen from... SQL, so I didn't think to document it as much as I probably should have.  When we add support for RANGE peers, we can remove the suppression and actually document it.

File: processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ */
+
+package org.apache.druid.query.operator.window;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+
+public class WindowFrame
```
[the remark this answers is not in the document]
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java:27]: DataFrame is a thing in R too...  I don't think it's far from what a `org.apache.druid.frame.Frame` is though, so maybe that's actually a good name?

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java (hunk 1 above)
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:77]: little secret: I needed them for myself :P

some code is "self documenting" some code is not, this code is not.

**Pull request merged**